### PR TITLE
Add module patching (mocking) feature to compiler

### PR DESCRIPTION
**Pull request by quantimnot**
Add NimScript and compiler APIs for patching modules.

### Procs `modulepaths.addModulePatch`, `nimscript.patchModule`

* Can patch any target module with a patch module.
* Can patch one of many modules with the same name and package.
* Patch modules can import the target module and then selectively import/export
and re-implement symbols.
* Target and patch module paths support normal path resolution so that a patch
can be in a distributed package.

Some uses:
* Mocking functionality for tests.
* Replacing or modifying modules that are used in third party packages.
* Instrumenting modules for debugging or profiling.
* Distributing patches in packages for others to use.

See tests/compilerfeatures/tpatchModule for an example.

### Compiler

These changes were to break cyclic import dependencies when hints and warnings
were add:

* Moved `options.findFile` into a new module named `searchpaths`.
* Moved `options.findModule` into `modulepaths`.

I searched GitHub and used https://sourcegraph.com to look for public repos that
are using those compiler APIs, but found none.

## Locations where paths are resolved during compile-time:

* `import`/`include` statements at `importer.myImportModule`
  - resolves the path with `findModule`
* `--import` at `commands.processSwitch`
  - resolves the path with `findModule`
* `--include` at `commands.processSwitch`
  - resolves the path with `findModule`
* `--link` at `extccomp.addExternalFileToLink`
  - does not resolve the path; uses it verbatim
  - **BUG: doesn't participate in patching when the `link` pragma does**
* `--compile` at `commands.processCompile`
  - resolves the path with `findFile`, or uses the path verbatim
  - **BUG: doesn't participate in patching when not resolved with `findFile`**
* `header` pragma at `pragmas.processLink`
  - resolves the path relative to the module, then with `findFile`
  - **BUG: doesn't participate in patching when the path resolves relative
  to the owning module**
* `link` pragma at `pragmas.processLink`
  - resolves the path relative to the module, then with `findFile`
  - **BUG: doesn't participate in patching when the path resolves relative
  to the owning module**
* `compile` pragma of form `{.compile: ("file.c", "$1.o").}` at `pragmas.processCompile`
  - **BUG: resolves the path relative to the module with `os.walkFiles`**
  - **BUG: supports wildcard patterns in path, but other pragma form doesn't**
  - **BUG: only supports relative paths**
  - **BUG: no path variable interpolation**
  - **BUG: doesn't participate in patching**
* `compile` pragma of form `{.compile: "file.c".}` at `pragmas.processCompile`
  - **BUG: doesn't participate in patching when the path resolves relative to the owning module**
  - resolves the path relative to the module, then with `findFile`
* `staticRead`/`slurp` at `vmdeps.opSlurp`
  - resolves the path relative to the module, then with `findFile`
  - **BUG: doesn't participate in patching when the path resolves relative to the owning module**
* `nimeval.evalScript`
  - resolves the script path with `findFile`
* docgen at `docgen.docgenFindFile`
  - resolves the path with `findFile`, then against the current dir
  - I'm not sure if the current dir is that of the module or of the invocation
* (when `RstGenerator` or `RstSharedState` initialized with not-nil `findFile` field)
  * RST `:file:` codeblock directive at `rst.dirCodeBlock`
    - resolves the path relative to the module/rst, then with `findFile`
    - **BUG: doesn't participate in patching when the path resolves relative
    to the owning module**
  * RST `:include:` directive at `rst.dirInclude`
    - resolves the path relative to the module/rst, then with `findFile`
    - **BUG: doesn't participate in patching when the path resolves relative
    to the owning module**
  * RST `:raw:` directive at `rst.dirRaw`
    - resolves the path relative to the module/rst, then with `findFile`
    - **BUG: doesn't participate in patching when the path resolves relative
    to the owning module**
  * RST `:image:` directive at `rst.dirImage`
     (This is an unfinished RST feature, that currently does nothing, but would most likely follow the pattern of other files read by the RST generator)
  * RST `:figure:` directive at `rst.dirFigure`
     (This is an unfinished RST feature, that currently does nothing, but would most likely follow the pattern of other files read by the RST generator)

## Other possibilities

* Overload `nimscript.patchFile` instead of making `options.patchModule`.
* **(Breaking for `nimscript.patchFile`)** Paths resolved relative to the current
module directory can also be patched.
* Deprecate `nimscript.patchFile`.

## TODO

* [x] move `modulepaths.findFile` ~~and `modulepaths.patchModule`~~ to `searchpaths`(?)
* ~~[ ] change `findFile` to try to resolve against a given absolute dir~~
* [x] test a cycle: target->patch->target
* [x] test a patched patch: target->patch0->patch1
* [x] test using a patch that is in a foreign package
* ~~[ ] test RST directives~~
* [x] test `nimscript.patchFile` for regressions
* [ ] test all locations where paths are resolved to ensure no side effects
* [ ] test performance impact
  - [ ] memory
  - [ ] CPU time
  - [ ] I/O
* [ ] changelog
* [ ] create PR for general path handling consistency and docs
* [ ] final cleanup
  - [ ] remove unneeded tests
  - [ ] remove dev notes

### CODE REVIEW

File: compiler/importer.nim [hunk 1]
```diff
@@ -239,8 +240,33 @@ proc importModuleAs(c: PContext; n: PNode, realModule: PSym, importHidden: bool)
...
+    #
+    # XXX: I'm not sure why an alias for `realModule` is created. See the
+    #      original comment below.
+    func moduleIdent(node: PNode): PIdent =
+      if node.kind == nkIdent:
+        # import module
+        node.ident
+      elif node.kind in nkStrKinds:
+        # import "/path/to/module.nim"
+        getIdent(c.cache, splitFile(node.strVal).name)
```
Comment (quantimnot) [inline compiler/importer.nim:262]: Should path substitution be performed here?
The only path sub that is a filename is `projectfile` and I don't see why that would
be patched.

File: compiler/modules.nim [hunk 2]
```diff
@@ -115,9 +116,38 @@ proc wantMainModule*(conf: ConfigRef) =
...
+        rawMessage(conf, warnCannotOpen, override)
+    else:
+      resolvedOverride = override
+    if resolvedTarget.len > 0 and resolvedOverride.len > 0:
+      conf.moduleOverridePaths[resolvedTarget] = resolvedOverride
+
 proc compileProject*(graph: ModuleGraph; projectFileIdx = InvalidFileIdx) =
   connectCallbacks(graph)
   let conf = graph.config
+  resolveModuleOverridePaths(conf)
```
Comment (quantimnot) [inline compiler/modules.nim]: The target and patch paths need resolved after all search paths are defined.
Is this the best place to do this?

**Comment by juancarlospaco** [comment]
How is the progress going?, this is very useful. 🙂

**Comment by quantimnot** [comment]
@juancarlospaco I have decided that this PR won't patch non-module paths like `patchFile`. This could change in the future though. So now, I just need to run this in CI to test my latest changes, fix anything broken, write a few more tests and then make any code review change requests.

I hope the idea is accepted. This may take a reviewer a good chunk of time, so it may be a while before the core devs can get to it.

**Comment by Araq** [comment]
> Was it your intention for nimscript.patchFile to not patch a path when it is relative to the script?

Not sure what that means but I think the answer is "no".

> Why did you have a note for ConfigRef.writelnHook that it couldn't be marked gcsafe because of Nimble? Is that resolved?

The CI will tell us.

> Does this affect IC?

I doubt it, IC is based on absolute paths, I think.

**Comment by quantimnot** [comment]
> Was it your intention for nimscript.patchFile to not patch a path when it is relative to the script?
> Not sure what that means but I think the answer is "no".

@Araq I opened #20053 with an example of what I mean.

Given the issues raised in #20053, should  I:
* deprecate `nimscript.patchFile(string,string,string)`
* change `nimscript.patchModule` to `nimscript.patchFile(string,string)` and allow it to patch statically read files other than modules
* change locations that resolve statically read paths to also patch paths that are resolved relative to the module

Also, I need to test error messages that include paths, but I was having a difficult time. I opened #19996 to address it. The current tests in this PR depend on that PR.

Thanks!

**Pull request closed without merging**